### PR TITLE
Feature: Effectful Auth Middleware

### LINKED ISSUE #936: Add `basicAuthZIO` helper in Auth Middleware
**Is your feature request related to a problem? Please describe.**
To enable working with ZIO inputs

**Describe the solution you'd like**
Need `basicAuthZIO` helper in Auth Middlewares. Let the parameter be ZIO values.

### PR DESCRIPTION
**Pull request by dinojohn**
closes #936

### CODE REVIEW

File: zio-http/src/main/scala/zhttp/http/middleware/Auth.scala [hunk 1]
```diff
@@ -18,6 +19,19 @@ private[zhttp] trait Auth {
       Headers(HttpHeaderNames.WWW_AUTHENTICATE, BasicSchemeName),
     )
 
+  /**
+   * Creates a middleware for basic authentication using an effectful
+   * verification function
+   */
+  final def basicAuthZIO[R, E](f: Header => ZIO[R, E, Boolean]): HttpMiddleware[R, E] =
```
Comment (dinojohn) [inline zio-http/src/main/scala/zhttp/http/middleware/Auth.scala]: Should I rewrite basicAuth and customAuth using basicAuthZIO and customAuthZIO?

File: zio-http/src/main/scala/zhttp/http/headers/HeaderGetters.scala [hunk 2]
```diff
@@ -74,7 +74,7 @@ trait HeaderGetters[+A] { self =>
   final def authorization: Option[CharSequence] =
     headerValue(HeaderNames.authorization)
 
-  final def basicAuthorizationCredentials: Option[Header] = {
+  final def basicAuthorizationCredentials: Option[(String, String)] = {
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/headers/HeaderGetters.scala]:
```suggestion
  case class Credentials(uname: String, upassword: String)

  final def basicAuthorizationCredentials: Option[Credentials] = {
```

File: zio-http/src/main/scala/zhttp/http/middleware/Auth.scala [hunk 3]
```diff
@@ -37,4 +51,18 @@ private[zhttp] trait Auth {
       _ => Middleware.identity,
       _ => Middleware.fromHttp(Http.status(Status.FORBIDDEN).addHeaders(responseHeaders)),
     )
+
+  /**
+   * Creates an authentication middleware that only allows authenticated
+   * requests to be passed on to the app using an effectful verification
+   * function.
+   */
+  final def customAuthZIO[R, E](
```
Comment (tusharmath) [inline zio-http/src/main/scala/zhttp/http/middleware/Auth.scala:52]: update implementation for `customAuthZIO`

File: zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala [hunk 4]
```diff
@@ -2,13 +2,17 @@ package zhttp.http.middleware
...
 import zhttp.internal.HttpAppTestExtensions
+import zio.UIO
 import zio.test.Assertion._
 import zio.test._
 
 object AuthSpec extends DefaultRunnableSpec with HttpAppTestExtensions {
-  private val basicHS    = Headers.basicAuthorizationHeader("user", "resu")
-  private val basicHF    = Headers.basicAuthorizationHeader("user", "user")
-  private val basicAuthM = Middleware.basicAuth { case (u, p) => p.toString.reverse == u }
+  private val basicHS: Headers                         = Headers.basicAuthorizationHeader("user", "resu")
```
Comment (tusharmath) [inline zio-http/src/test/scala/zhttp/http/middleware/AuthSpec.scala]:
```suggestion
  private val basicHS: Headers                         = Headers.basicAuthorizationHeader(Credentials("user", "resu"))
```

**Review by tusharmath — APPROVED** [review]

**Pull request merged**